test/metric_value_benchmark/statsd_test.go

```diff
@@ -44,62 +68,123 @@ func (t *StatsdTestRunner) GetAgentConfigFileName() string {
-	// EC2 Image Builder creates a bash script that sends statsd format to cwagent at port 8125
-	// The bash script is at /etc/statsd.sh
-	//    for times in  {1..3}
-	//    do
-	//      echo "statsd.counter:1|c" | nc -w 1 -u 127.0.0.1 8125
-	//      sleep 60
-	//    done
-	startStatsdCommand := []string{
-		"sudo bash /etc/statsd.sh",
+	// For each test run we want a unique metric name.
+	// Just in case 2 different people run this test in parallel, using the
+	// same AWS account...
+	// Populate a temp map with the new metric names.
+	suffix := fmt.Sprint(time.Now().UnixNano())
+	newMap := map[string]metricInfo{}
+	for k, v := range metricMap {
+		newKey := k + suffix
+		newMap[newKey] = v
 	}
-
-	return common.RunCommands(startStatsdCommand)
+	// Replace global map with the temp one.
+	metricMap = newMap
+
+	// Send unique metrics each second.
+	// Expect agent to collect every 5 seconds.
+	// Expect agent to aggregate collections into 30 second buckets.
+	go sendStatsd()
+	return nil
 }
 
 func (t *StatsdTestRunner) GetMeasuredMetrics() []string {
-	return []string{"statsd_counter"}
+	keys := make([]string, 0, len(metricMap))
+	for k := range metricMap {
+		keys = append(keys, k)
+	}
+	return keys
 }
 
 func (t *StatsdTestRunner) validateStatsdMetric(metricName string) status.TestResult {
 	testResult := status.TestResult{
 		Name:   metricName,
 		Status: status.FAILED,
 	}
-
-	dims, failed := t.DimensionFactory.GetDimensions([]dimension.Instruction{
+	// Get the metric info for the current metricName.
+	// Just assume it exists.
+	metricInfo := metricMap[metricName]
+	// Populate the list of expected dimensions.
+	instructions := []dimension.Instruction{
 		{
 			Key:   "InstanceId",
 			Value: dimension.UnknownDimensionValue(),
 		},
 		{
+			// CWA adds this metric_type dimension.
 			Key:   "metric_type",
-			Value: dimension.ExpectedDimensionValue{aws.String("counter")},
+			Value: dimension.ExpectedDimensionValue{Value: &metricInfo.metricType},
 		},
-	})
-
+	}
+	//
+	for _, d := range metricInfo.dimensions {
+		instructions = append(instructions, dimension.Instruction{
+			Key: d[0],
+			Value: dimension.ExpectedDimensionValue{Value: aws.String(d[1])},
+		})
+	}
+	dims, failed := t.DimensionFactory.GetDimensions(instructions)
 	if len(failed) > 0 {
 		return testResult
 	}
-
 	fetcher := metric.MetricValueFetcher{}
-	values, err := fetcher.Fetch(namespace, metricName, dims, metric.AVERAGE, test_runner.HighResolutionStatPeriod)
+	// Namespace must match the JSON config.
+	values, err := fetcher.Fetch("statsd_test", metricName, dims, metric.AVERAGE,
+		test_runner.HighResolutionStatPeriod)
 	if err != nil {
 		return testResult
 	}
-
+	// Aggregation interval is 30 seconds, so expect 2 * runTimeInMinutes.
+	numExpected := 2 * int(t.GetAgentRunDuration().Minutes())
+	if numExpected != len(values) {
+		log.Printf("fail: expected %v data points, got %v",
+			numExpected, len(values))
+	}
```

Comment:
I was pleasantly surprised to see this work.
I am worried that the CW web service could be slow though and that we won't see all the data points show up. `Fetch()` doesn't delay or retry... So maybe I should allow for 1 missing data point?

Reply:
Eventually this did fail for me, so I am going to change the validation

Reply:
Reran the latest test code 10 times without failure:
```
[ec2-user@ip-172-31-17-153 amazon-cloudwatch-agent-test]$ go test -failfast -p=1 -count=10 -timeout=60m ./test/metric_value_benchmark/ -computeType=EC2
ok  	github.com/aws/amazon-cloudwatch-agent-test/test/metric_value_benchmark	1810.983s
```

Reply:
Allowing for 2 missing data points makes sense to me